### PR TITLE
Add get_map_objects_min_distance_meters logic

### PR DESCRIPTION
**Pull request by hunterjm**
Implement map objects and interaction distance restrictions via the haversine formula.  Also updated the download settings to be saved on every API call in order to update the restrictions.

The haversine formula is in Utils so it can be used to implement the encounter restrictions also found in the download settings response.  Since the pokemon lat,lng isn't passed into the client, however, that restriction will need to be implemented by the consumer.

### CODE REVIEW

File: pogobuf/pogobuf.client.js [hunk 1]
```diff
@@ -15,7 +15,10 @@ const RequestType = POGOProtos.Networking.Requests.RequestType,
     Responses = POGOProtos.Networking.Responses;
 
 const INITIAL_ENDPOINT = 'https://pgorelease.nianticlabs.com/plfe/rpc';
-const DEFAULT_MAP_OBJECTS_DELAY = 5;
+const DEFAULT_MAP_OBJECTS_DELAY = 30;
+const DEFAULT_MAP_OBJECTS_DISTANCE = 10;
+const DEFAULT_FORT_INTERACTION_RANGE = 40;
+const DEFAULT_FORT_INTERACTION_RANGE_FAR = 1000;
```
Comment (ghost) [inline pogobuf/pogobuf.client.js]: I think these should be moved directly to the `internal stuff` section. (to stop clutter)

File: pogobuf/pogobuf.client.js [hunk 2]
```diff
@@ -804,7 +826,11 @@ function Client() {
 
     this.maxTries = 5;
     this.mapObjectsThrottlingEnabled = true;
-    this.mapObjectsMinDelay = DEFAULT_MAP_OBJECTS_DELAY * 1000;
+    this.mapObjectsMaxDelay = DEFAULT_MAP_OBJECTS_DELAY * 1000;
+    this.mapObjectsMinDistance = DEFAULT_MAP_OBJECTS_DISTANCE;
+    this.interactionThrottlingEnabled = true;
```
Comment (ghost) [inline pogobuf/pogobuf.client.js]: Probably best to leave this to false to not break any applications (if backwards compatibility is a thing). If someone _really_ wants to have this kind of throttling enabled they should enable it themselves.

File: pogobuf/pogobuf.client.js [hunk 3]
```diff
@@ -953,23 +979,40 @@ function Client() {
...
         if (requests.some(r => r.type === RequestType.GET_MAP_OBJECTS)) {
+
+            // Check distance traveled to see if it is past refresh distance
+            var refreshRequired = !self.withinInteractionRange(
+                self.lastMapObjectsLatitude,
+                self.lastMapObjectsLongitude,
+                self.mapObjectsMinDistance);
             var now = new Date().getTime(),
-                delayNeeded = self.lastMapObjectsCall + self.mapObjectsMinDelay - now;
+                delayNeeded = self.lastMapObjectsCall + self.mapObjectsMaxDelay - now;
```
Comment (cyraxx) [inline pogobuf/pogobuf.client.js:962]: Can you explain why the change from min to max delay?

File: pogobuf/pogobuf.utils.js [hunk 4]
```diff
@@ -248,6 +248,30 @@ module.exports = {
     },
 
     /**
+     * Utility method to get the distance in meters between two locations
+     * @param {number} lat1 - origin latitude
+     * @param {number} lng1 - origin longitude
+     * @param {number} lat2 - destination latitude
+     * @param {number} lng2 - destination longitude
```
Comment (ghost) [inline pogobuf/pogobuf.utils.js]: These could be named like:
- originLatitude
- originLongitude
- destinationLatitude
- destinationLongitude

or destLng etc.

File: pogobuf/pogobuf.client.js (hunk 2 above)
Comment (hunterjm) [reply pogobuf/pogobuf.client.js]: Correct me if I'm wrong, but I think the API returns a failure if you are too far away anyway, so this would be preventing an unnecessary API call.

File: pogobuf/pogobuf.client.js (hunk 3 above)
Comment (hunterjm) [reply pogobuf/pogobuf.client.js:962]: This would better mimic the app, which I think is what you were going for here anyway.  When standing still, my logs show a map objects request every 30 seconds. It's when your moving that it becomes more frequent, which is handled by the refresh required logic.

File: pogobuf/pogobuf.utils.js (hunk 4 above)
Comment (hunterjm) [reply pogobuf/pogobuf.utils.js]: They could be, but origin/destination are not really a part of the formula. It just calculates distance between two points, which could technically be flipped to destination/origin with the same result.

File: pogobuf/pogobuf.client.js (hunk 2 above)
Comment (ghost) [reply pogobuf/pogobuf.client.js]: If I remember correctly, you can do some actions outside told max distance, but it increases the possibility of getting the account banned.

File: pogobuf/pogobuf.utils.js (hunk 4 above)
Comment (ghost) [reply pogobuf/pogobuf.utils.js]: Then you should change your JSDoc to say "latitude 1" "longitude 1" etc.

File: pogobuf/pogobuf.client.js (hunk 3 above)
Comment (ghost) [reply pogobuf/pogobuf.client.js:962]: Personally I think the library should not throttle the requests too much – it is the library user's decision whether (s)he wants to replicate the real application. Many scanners would break with this change, too.

Comment (hunterjm) [reply pogobuf/pogobuf.client.js:962]: @joelfi: I would agree that the library shouldn't throttle too much, but maybe **all** throttling should be disabled by default then?  While I agree that we shouldn't limit the abilities of people who want to use the API in a non-standard way, I still think the client library is the correct place for that type of logic.

I'm lost as to how this would break scanners though.  I'm a scanner.  I move my location when I scan.  The center cell of my new scan location is likely much further than 10 meters than the center cell of my previous scan location since I want to cover as much ground as possible with as little overlap as possible.

The old implementation of this delay logic would require me to disable throttling or wait 10 seconds before I can scan again.  The new implementation of this logic would have the same impact as if throttling was disabled, because in all likelihood the scanner position has moved more than 10 meters.

This is also similar to the logic @AeonLucid uses in his [POGOLib](https://github.com/AeonLucid/POGOLib#heartbeats)

File: pogobuf/pogobuf.client.js (hunk 1 above)
Comment (hunterjm) [reply pogobuf/pogobuf.client.js]: Where is the `internal stuff` section?

**Comment by MakuraYami** [comment]
There's quite some controversial over this PR but i think everyone is forgetting 2.0 branch. Why are people ignoring it and continuing to develop here? 

Back on topic though, cyraxx has told me multiple times this project is purely an API. How it's used is up to the user. I think this PR goes against that idea entirely.

File: pogobuf/pogobuf.client.js (hunk 1 above)
Comment (ghost) [reply pogobuf/pogobuf.client.js]: [Line 815 in fork `hunterjm:feature/distance`](https://github.com/hunterjm/pogobuf/blob/edb2cc47eed3be04e201b1fe4d4a824e3d269c14/pogobuf/pogobuf.client.js#L815).

**Comment by hunterjm** [comment]
@MakuraYami: The 2.0 branch is a month old with little to no progress according to the issue.  Do you have a timeline for release yet?  If you think this belongs there, that is fine, I can move it.

I've removed the interaction range portion of the PR to put to rest some of the controversy since the interaction distance logic is handled easily enough outside the client.

Now, if this is indeed intended to purely be an API client with no additional logic, I would be happy to update the PR to pull out the existing throttling logic based on `get_map_objects_min_refresh_seconds`.  If, however, the community decides that map object throttling is valuable to have inside the client so scanners can have a `setInterval` and forget approach, than I believe that the distance logic for maps object does nothing but enhance that.

**Comment by hunterjm** [comment]
I just read through https://github.com/cyraxx/pogobuf/issues/54 which is where the throttling was added initially to prevent errors from Niantic.  I will do some more testing on my end with the client, but I believe that the requests can be made faster than the this library's 10 second minimum delay so long as it is not being done in the same location and you are moving at a rate of speed faster than 1 meter per second (If your player position is in the same location, the difference between the min 10 seconds and max 30 seconds wouldn't really matter in the grand scheme of things.)  If that is the case, scanners that are using this API can be made more efficient by changing the default throttling to include the distance limitations.

**Comment by cyraxx** [comment]
What I usually say is that the purpose of the library is to enable you to call API methods while preventing you from getting error messages as much as possible. That's why we have the map objects delay and the retry logic: So you can make your map objects calls without having to worry about rate limiting errors. However, the library still allows you to make API calls that don't make sense or will get you blocked, and it does not implement any "human behavior simulation". So if you make one call that puts you in Europe and then another one right afterwards that puts you in Australia, it will let you do that. It doesn't stop you from always getting 100% perfect Pokémon catches either.

The logic you provided is kind of an edge case. If you do the testing you mentioned and it results in enabling us to decrease delay times while still being as neutral as possible, I'm not against including it. And don't get me wrong, I very much appreciate your contribution either way.

@MakuraYami is also right though: We should focus on getting the 2.0 branch finished and migrate the library to 2.0 instead of keeping the two separate versions. I will make 2.0 my priority before adding major new features that would then only have to be ported to the other branch.

**Comment by cyraxx** [comment]
Closing all issues as this repository is no longer being maintained. See README for further information.
